### PR TITLE
Ignore internal TableAlreadyExistsException in Iceberg

### PR DESCRIPTION
**Pull request by pajaks**
## Description
Ignore internal TableAlreadyExistsException in CREATE TABLE & register_table on Iceberg connector
Fixes https://github.com/trinodb/trino/issues/18308

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/AbstractMetastoreTableOperations.java [hunk 1]
```diff
@@ -105,6 +106,7 @@ protected final void commitNewTable(TableMetadata metadata)
                 .withStorage(storage -> storage.setStorageFormat(ICEBERG_METASTORE_STORAGE_FORMAT))
                 // This is a must-have property for the EXTERNAL_TABLE table type
                 .setParameter("EXTERNAL", "TRUE")
+                .setParameter(PRESTO_QUERY_ID_NAME, session.getQueryId())
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/AbstractMetastoreTableOperations.java:109]: I'm not sure whether this is necessary.

Take a look whether `metadata.currentSnapshot().summary()` contains already `trino_query_id` .

In this way we could apply the fix potentially for all the implementations instead of doing it only for Hive metastore.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergCreateTableInternalRetry.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    @Test
+    public void testCreateTableAsSelectInternalRetry()
+    {
+        assertQuerySucceeds("CREATE TABLE test_ctas_internal_retry AS SELECT 1 a");
+        assertQuery("SHOW TABLES LIKE 'test_ctas_internal_retry'", "VALUES 'test_ctas_internal_retry'");
+    }
+
+    @Test
+    public void testRegisterTableInternalRetry()
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergCreateTableInternalRetry.java:122]: Is this test related to this commit?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/AbstractMetastoreTableOperations.java (hunk 1 above)
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/AbstractMetastoreTableOperations.java:109]: is PRESTO_QUERY_ID_NAME parameter being updated when committing a change to an existing table?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/AbstractMetastoreTableOperations.java [hunk 3]
```diff
@@ -113,12 +113,32 @@ protected final void commitNewTable(TableMetadata metadata)
...
             metastore.createTable(table, privileges);
         }
-        catch (SchemaNotFoundException
-               | TableAlreadyExistsException e) {
+        catch (SchemaNotFoundException e) {
             // clean up metadata files corresponding to the current transaction
             fileIo.deleteFile(newMetadataLocation);
             throw e;
         }
+        catch (TableAlreadyExistsException e) {
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/AbstractMetastoreTableOperations.java:121]: potential follow-up: Please add a similar handling in `io.trino.plugin.iceberg.catalog.glue.GlueIcebergTableOperations#commitNewTable`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java [hunk 4]
```diff
@@ -21,6 +21,7 @@
 import io.trino.filesystem.TrinoFileSystemFactory;
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java:21]: nit: Consider changing the commit comment "Ignore internal TableAlreadyExistsException in Iceberg register table procedure" -> "Ignore internal TableAlreadyExistsException in register_table procedure"

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java [hunk 5]
```diff
@@ -317,11 +318,34 @@ public void registerTable(ConnectorSession session, SchemaTableName schemaTableN
...
                 // This is a must-have property for the EXTERNAL_TABLE table type
                 .setParameter("EXTERNAL", "TRUE")
+                .setParameter(PRESTO_QUERY_ID_NAME, session.getQueryId())
                 .setParameter(TABLE_TYPE_PROP, ICEBERG_TABLE_TYPE_VALUE.toUpperCase(ENGLISH))
                 .setParameter(METADATA_LOCATION_PROP, tableMetadata.metadataFileLocation());
 
         PrincipalPrivileges privileges = owner.map(MetastoreUtil::buildInitialPrivilegeSet).orElse(NO_PRIVILEGES);
-        metastore.createTable(builder.build(), privileges);
+        try {
+            metastore.createTable(builder.build(), privileges);
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java:327]: potential follow-up: Add this functionality for TrinoGlueCatalog as well.

**Review by findinpath — APPROVED** [review]
Please consider adding handling for Glue either in the scope of this PR or in a follow-up PR.

**Review by findinpath — APPROVED** [review]

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
/test-with-secrets sha=4cb09a0f7c64aa6f8d66a6d164df24238456778a

**Comment by pajaks** [comment]
CI fialures:
-  `ci / pt (default, suite-1, )` - `io.trino.tests.product.hive.TestHiveTransactionalTable.testReadFullAcid` - seems unrelated
- `ci / maven-checks (21)` seems unrelated, especially with_secrets passed

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
`ci / pt (default, suite-1, ) ` hit #16315
